### DISCUSSION TITLE
I/O attempted in an unexpected place with asyncio extension

### DISCUSSION THREAD

**Discussion by miguelgrinberg** [sqlalchemy/sqlalchemy#9242] [Usage Questions] [answered]
Hi Mike & Team. I'm working on a little project with the asyncio extension (SQLAlchemy 2.0.1) and have a problem I was hoping you can help me understand.

I'm going to share a single file project that you can run if you need to. I've made it as short as I could.

```python
import asyncio
from sqlalchemy import ForeignKey, select
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker
from sqlalchemy.orm import declarative_base, Mapped, mapped_column, relationship

engine = create_async_engine('sqlite+aiosqlite://')
Model = declarative_base()
Session = async_sessionmaker(engine, expire_on_commit=False)

class Order(Model):
    __tablename__ = 'orders'
    id: Mapped[int] = mapped_column(primary_key=True)
    order_items: Mapped[list['OrderItem']] = relationship(lazy='joined')

class OrderItem(Model):
    __tablename__ = 'order_items'
    order_id: Mapped[int] = mapped_column(ForeignKey('orders.id'), primary_key=True)

async def main():
    async with engine.begin() as conn:
        await conn.run_sync(Model.metadata.create_all)
    async with Session() as session:
        async with session.begin():
            o = Order()
            session.add(o)
            # <-- some database queries here appear to expire the `o` object
            o.order_items.append(OrderItem())

asyncio.run(main())
```

The above code works great. But when I add certain database queries right after the `session.add(o)` it appears the `o` object is marked as expired, so then the `o.order_items` attribute access triggers a background load, which I think makes no sense because this is a brand new object.

So for example, the error appears if I add the following query in the line where the comment is:

```python
            await session.scalar(select(OrderItem).limit(1))
```

But interestingly, this other query does not cause any errors:

```python
            await session.scalar(select(Order).limit(1))
```

Many other queries for other entities that I have in my real project also trigger the error. Also if I move the offending query above the `session.add(o)` call, then everything works. Is there anything obvious here that I'm missing? Thanks!

**Comment by CaselIT** [comment]
Hi,

I think we had something similar reported, I'll try to look for it.
Does adding to the relationship `default_factory=list` help?

**Reply by miguelgrinberg** [reply]
Is `default_factory` an option for a `relationship()` declaration? Seems it is related to the dataclass integration, which I'm not using. I get an error when I add it:

```text
sqlalchemy.exc.ArgumentError: Attribute 'order_items' on class <class '__main__.Order'> includes dataclasses argument(s): 'default_factory' but class does not specify SQLAlchemy native dataclass configuration.
```

**Reply by CaselIT** [reply]
I think the something similar is this https://github.com/sqlalchemy/sqlalchemy/discussions/8913

> Is default_factory an option for a relationship() declaration

yes, of course. Sorry I kind of assumed dataclass, since 90% of issues use that with v2. I don't remember if there is something similar when not using dataclass. Maybe the `validates` decorator could be used (or a custom `__init__`)

**Reply by miguelgrinberg** [reply]
Yes, I think #8913 is related, thanks. I was able to reproduce the error when adding a flush instead of a query as in that discussion. And the error goes away after initializing the relationship with an empty list as suggested by Mike there.

But this feels like a hack, don't you think? I can access the relationship before a flush and it gets automatically transformed into an empty list on the fly. Just because I flushed, now this behavior disappears and the relationship is now considered expired. At the very least, I think this deserves to be explained in https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession?

**Reply by zzzeek** [reply]
the relationship was not expired, it was just never there in the first place.    so you've flushed this `Order` object and not told it anything about `order.order_items`, the ORM has to assume maybe you've populated the database in some other way.

surprised it's not in that doc you mention though, yes it should be mentioned there with an example .

**Reply by miguelgrinberg** [reply]
Well, I didn't flush anything. I just ran a query, and SQLAlchemy decided to flush on its own. That is what made this even more confusing to me. Not all queries that I issued caused an implicit flush, but many did. I still haven't figured out exactly what the rules are, i.e. when SQLAlchemy needs to flush as a result of a select query and when it doesn't. As you can see in the code I shared above, I was using a completely empty database, so nothing was being loaded with these queries, yet a flush sometimes still happened.

I totally get that this seems logical to you, being familiar with the internals, but the behavior does seem very inconsistent when you look at it from the outside. I can add to a relationship that I haven't explicitly initialized, as long as no flush has occurred. After a flush, SQLAlchemy believes it needs to refresh the uninitialized relationship, but it didn't care before the flush. And flushes occur outside of the developer's control at least that is my perception, not having a yet a full understanding of when the flush has to happen.

> the ORM has to assume maybe you've populated the database in some other way.

This is a new order object that has not been committed yet. How could there be any related items added from another process?

**Reply by zzzeek** [reply]
> Well, I didn't flush anything. I just ran a query, and SQLAlchemy decided to flush on its own. That is what made this even more confusing to me. 

well that's autoflush, which you can also turn off.

> Not all queries that I issued caused an implicit flush, but many did. 

all queries trigger autoflush when it is enabled.

your query here:

    await session.scalar(select(Order).limit(1))

triggered autoflush.  Why no error?  Because after it flushed the Order, it then SELECTed it, and as you have lazy='joined' for order.order_items, it populated that previously un-populated attribute for you.

> I totally get that this seems logical to you, being familiar with the internals, but the behavior does seem very inconsistent when you look at it from the outside. 

All of these issues, pretty much in their entirety, are because ORM models have this behavior of collections being lazy-loaded, which is not compatible with asyncio.  A real asyncio ORM would have no lazyloading whatsoever and it probably would not even have `obj.collection` being readable.    That's why in SQLAlchemy 2.0 we have a new relationship system called [write only relationships](https://docs.sqlalchemy.org/en/20/orm/large_collections.html) that establishes a pattern for loading relationships that does not use object attributes at all, and always requires an [explicit query](https://docs.sqlalchemy.org/en/20/orm/large_collections.html#querying-items).  This is really what asyncio is all about, so I will likely move to advertise this pattern a lot more directly on the asyncio page.    

So I would say if you really want to write asyncio and do it fully in the spirit of what asyncio is about, that is, all IO is explicit at all times, I would start using write only relationships in all cases.  all of these problems will go away.


> > the ORM has to assume maybe you've populated the database in some other way.
> 
> This is a new order object that has not been committed yet. How could there be any related items added from another process?

it was autoflushed and people definitely do add objects to the database by indicating their foreign key identity, like this:

    o = Order(id=1)
    session.add(o)
    session.add_all([OrderItem(order_id=1), OrderItem(order_id=1), OrderItem(order_id=1)])
    session.flush()

if we assumed that o.order_items was an empty list, and positively set the empty list there, now it's got the wrong data.

**Reply by zzzeek** [reply]
here is the program using write only relationships.  it will never raise an IO error on collection access:

```py
import asyncio
from sqlalchemy import ForeignKey, select
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker
from sqlalchemy.orm import declarative_base, Mapped, mapped_column, relationship, WriteOnlyMapped

engine = create_async_engine('sqlite+aiosqlite://')
Model = declarative_base()
Session = async_sessionmaker(engine, expire_on_commit=False)

class Order(Model):
    __tablename__ = 'orders'
    id: Mapped[int] = mapped_column(primary_key=True)
    order_items: WriteOnlyMapped['OrderItem'] = relationship(lazy='write_only')

class OrderItem(Model):
    __tablename__ = 'order_items'
    order_id: Mapped[int] = mapped_column(ForeignKey('orders.id'), primary_key=True)

async def main():
    async with engine.begin() as conn:
        await conn.run_sync(Model.metadata.create_all)
    async with Session() as session:
        async with session.begin():
            o = Order()
            session.add(o)
            o.order_items.add(OrderItem())

    async with Session() as session:
        o = (await session.scalars(select(Order))).one()

        # prints <sqlalchemy.orm.writeonly.WriteOnlyCollection object at 0x7f37f666ae30>
        print(o.order_items)

        # raises an informative error
        # print(list(o.order_items))

        # actually view the items in the collection
        print((await session.scalars(o.order_items.select())).all())


asyncio.run(main())
```

**Reply by miguelgrinberg** [reply]
Thanks for the detailed explanation, Mike. This is starting to make some sense. But I still have a lot of questions.

The main thing that still feels odd to me is that I have created this Order instance, so it is brand new. If I don't do anything that triggers the autoflush, then there is no implicit I/O when I access the relationship the first itme. If I do something that requires flushing, or if I flush explicitly, then only at that point SQLAlchemy starts worrying that this relationship might be out of date and attempts to autoload it when the attribute is accessed. But we are talking about a new object. Yes, it has now been flushed, but it cannot be seen outside of this one session yet, so why is there a need to autoload it? I guess SQLAlchemy lost track that this is a new entity after the flush? [Edit: I think I now understand the point you made wrt hardcoded foreign keys that bypass the relationship object, so I see the need to load the relationship, but only if the loader is lazy, I guess]

I set the relationship to joined to prevent ever having an instance with an expired or uninitialized relationship, so it seems autoloads on "new objects that have been flushed" are a sort of hole in the recommended measures to coach SQLAlchemy into never needing to autoload a relationship. Initializing the relationship to [] does it, but if I specify joined or any other eager loader, why isn't the relationship initialized that way automatically? Why would I want that relationship to be left uninitialized when I clearly indicated I don't want to lazy load it?

I'm also well aware of the write_only loader, in fact you discussed with me your plan to add it last year in another discussion here. My impression, though, was that any relationship loader that is not select or dynamic was safe to use under asyncio. For this one relationship I wanted to avoid having to deal with with the write_only semantics, as it is a short relationship. I thought the use case for write_only is to be able to customize the query that gets the related list of objects with pagination, sorting, additional filters, etc. none of which I need for this relationship. Write only makes sense for long relationships, I think, but it is inconvenient for short/quick ones. This was also the case for the old dynamic loader, at least in my view.

**Reply by zzzeek** [reply]
> The main thing that still feels odd to me is that I have created this `Order` instance, so it is brand new. If I don't do anything that triggers the autoflush, then there is no implicit I/O when I access the relationship the first itme. If I do something that requires flushing, or if I flush explicitly, then only at that point SQLAlchemy starts worrying that this relationship might be out of date and attempts to autoload it when the attribute is accessed. But we are talking about a new object. Yes, it has now been flushed, but it cannot be seen outside of this one session yet, so why is there a need to autoload it?

in the example I showed, the OrderItem objects are added by id.  There's other possibilities also, like if you had a relationship with a "secondary" table that is also mapped to an association object, or if multiple relationships refer to the same thing, etc..

the only other possibility SQLAlchemy could do here is, it knows the object was flushed, it's in the database, it knows nothing at all about `order.order_items`, and it just sticks an empty list there, assuming that's what it must be.   This would just not work for the full expanse of SQLAlchemy applications out there because the ORM is guessing.    The only way to really handle this is that the object gets this empty list set on it upon construction.  

The most likely way people will be doing this pattern automatically a lot more is in the next few major releases is the dataclasses thing is also going to be used much more widely.  that will also fix the "default list on construction" issue.       

> I guess SQLAlchemy lost track that this is a new entity after the flush? [Edit: I think I now understand the point you made wrt hardcoded foreign keys that bypass the relationship object, so I see the need to load the relationship, but only if the loader is lazy, I guess]

The ORM tracks all the objects you give it but it does not automatically figure out every possible linkage of objects based on just some of them.  If you made these OrderItems point to Order via backref for example, there are other technical reasons in that case the ORM doesn't necessarily know that Order.order_items is getting those objects (backrefs don't populate collections that were not set up at all.  this is a big performance concern with a lot of history behind it).

I have to stress that this whole thing is a "problem" only because of asyncio, which we only added a few years ago in an advancement that IMO was nothing short of miraculous that it was even possible.   This issue is inconvenient but it's not insurmountable.


> I set the relationship to `joined` to prevent ever having an instance with an expired or uninitialized relationship, so it seems autoloads on "new objects that have been flushed" are a sort of hole in the recommended measures to coach SQLAlchemy into never needing to autoload a relationship. 

yeah my personal preference is to set relationships to lazy="raise" and then when querying, use the most appropraite eager loader. The model is always, "image there was no ORM, because ORMs are terrible, and we 'just want to write SQL'" like everyone always says.  So OK, we'd need to know here to load Order plus order_items, let's do the selectinload() on that (I also always prefer selectinload() for collections).

> Initializing the relationship to `[]` does it, but if I specify `joined` or any other eager loader, why isn't the relationship initialized that way automatically? 

eager loading doesnt imply anything about how the constructor works

> Why would I want that relationship to be left uninitialized when I clearly indicated I don't want to lazy load it?

that's not indicated at all?  you could load with the lazyload() option later on, or you could be populating objects into that collection via other means, i.e. other relationships, other objects with particular FK values, etc.  this is not a made up use case.  I've been tuning these operations for 16 years.   it's only asyncio that's making it temporarily more difficult

> I'm also well aware of the `write_only` loader, in fact you discussed with me your plan to add it last year in another discussion here. My impression, though, was that any relationship loader that is not `select` or `dynamic` was safe to use under asyncio. 

not at all, any relationship can be expired and it will lazy load when accessed unless you default it to lazy='raise'

> For this one relationship I wanted to avoid having to deal with with the `write_only` semantics, as it is a short relationship. I thought the use case for `write_only` is to be able to customize the query that gets the related list of objects with pagination, sorting, additional filters, etc. none of which I need for this relationship. Write only makes sense for long relationships, I think, but it is inconvenient for short/quick ones. 

OK ...but...why asyncio then? See this is the thing.  **asyncio is explicitly inconvenient**.  It's designed such that it forces you to add "async" "await" **everywhere**, when you could just as easily use gevent and get the **identical** runtime behavior with **no change to code whatsoever**.   Why use asyncio but then verbosity is less convenient?


you could also use asyncio with SQLAlchemy without any of these issues by using the `run_sync()` approach.  I dont expect people to do that either because people really love asyncio.  thats great, but object.foobar being magically available no matter what without any SQL is just not doable.

**Reply by zzzeek** [reply]
@CaselIT what do you think of making the dataclasses "default" and "default_factory" work for non-dataclasses?   maybe we can build it into the declarative default constructor.

**Reply by CaselIT** [reply]
@zzzeek I think you modified the message by error

**Reply by CaselIT** [reply]
Let me try fixing

**Reply by zzzeek** [reply]
thanks, sorry @miguelgrinberg i make that mistake a lot

**Reply by CaselIT** [reply]
Should be fixed now. GitHub it's great that keeps the older revisions



> @CaselIT what do you think of making the dataclasses "default" and "default_factory" work for non-dataclasses? maybe we can build it into the declarative default constructor.

I was thinking about it before, when I wrongly suggested the default_factory. It will complicate the default init a bit though

**Reply by zzzeek** [reply]
> Should be fixed now. GitHub it's great that keeps the older revisions

thanks, sorry, I'm sure you're noticing im making more old person mistakes...

**Reply by miguelgrinberg** [reply]
> yeah my personal preference is to set relationships to lazy="raise" and then when querying, use the most appropraite eager loader.

Would that help in this case though? Instead of the missing greenlet error I would get the error raised by this loader. The app would still fail. It seems there is no way to get a working relationship that uses list semantics but is configured to absolutely under no circumstance ever lazy load. That is, I think, what's missing here.

Basically I want to take responsibility for keeping my objects up to date. And I don't want to lose the ability to use my eager loaders. I think the current release is almost there, because this works great for all cases except for new objects that are flushed.

> not at all, any relationship can be expired and it will lazy load when accessed

Yes, but you have already documented very clearly that expire shouldn't be used, so that is not a problem in my view. The problem we are discussing here has no documented workaround yet. Should the docs ask people to always initialize their relationships as empty lists? I guess that would work, but to me that feels strange.

> OK ...but...why asyncio then?

My personal reason is that I'm working on a SQLAlchemy 2.0 tutorial, and I would like to cover the async support. I'm just trying to come up with a viable way to use it in real-world projects that I can demonstrate. I have used gevent and eventlet extensively and I agree 100% that they are rock solid for async development, but nobody wants to write code that way anymore.

> you could also use asyncio with SQLAlchemy without any of these issues by using the run_sync() approach.

The solution you came up to mix sync and async code is super clever and my mind exploded when I learned it from you, but I don't think you can expect async developers will want to use `run_sync()` in their code.

**Reply by zzzeek** [reply]
> > yeah my personal preference is to set relationships to lazy="raise" and then when querying, use the most appropraite eager loader.
> 
> Would that help in this case though? Instead of the missing greenlet error I would get the error raised by this loader. The app would still fail. 

well it would behave more consistently at least, I think the inconsistency you see is part of the problem you had.

> It seems there is no way to get a working relationship that uses list semantics but is configured to absolutely under no circumstance ever lazy load. That is, I think, what's missing here.

lazy="raise" means it will never lazyload.   I think what you are asking for is a list that can in 100% of cases **know** what's supposed to be inside of it, even if it means guessing, so that upon accessing it, the list always has contents inside of it and never needs to query the database.   That is simply not feasible.   Basic example, someone emits a query to load the parent object, and they explicitly **do not** indicate any eager loader, because they don't actually want the additional SQL.    The collection will be unloaded.  Later, someone edits the code that wants to access the collection, lazy="raise" and it raises.  The programmer **has** to go to that original query and alter it.   Having the unloaded collection simply return "blank" would be a major mistake.  it has to raise without having instructions as to how that collection should be populated.   in your example program, you have not given it that information.  I know you keep refuting this, but that's how it works, unless the constructor is changed, of which there are **many** ways to achieve that on either a manual or an automated basis.  


> 
> Basically I want to take responsibility for keeping my objects up to date. And I don't want to lose the ability to use my eager loaders. I think the current release is almost there, because this works great for all cases except for new objects that are flushed.

it works for new objects if you set up a default list constructor. this works if you use [declarative dataclass mappings](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping), or even just set up a custom constructor that adds a default collection.   this is not a hard problem and I dont know that it's worth pages and pages of back and forth for something we are already [proposing, right here to make more available by default](https://github.com/sqlalchemy/sqlalchemy/discussions/9242#discussioncomment-4877302).   

> 
> > not at all, any relationship can be expired and it will lazy load when accessed
> 
> Yes, but you have already documented very clearly that expire shouldn't be used, so that is not a problem in my view. The problem we are discussing here has no documented workaround yet. 

I dont know why you are saying that?   Literally every reply in this thread includes the workaround that was first noted in #8913 and as you requested is [now on the live site in the docs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession), see bullet number four, it's up, with bold text....


> Should the docs ask people to always initialize their relationships as empty lists? I guess that would work, but to me that feels strange.

but...it's a workaround?   as I have mentioned several times, we can expand upon the dataclasses feature to make this workaround more convenient, or just make it more of a default that people should use dataclasses mappings (which are completely brand new) up front, once this part of the library matures a bit.

> 
> > OK ...but...why asyncio then?
> 
> My personal reason is that I'm working on a SQLAlchemy 2.0 tutorial, and I would like to cover the async support. I'm just trying to come up with a viable way to use it in real-world projects that I can demonstrate. I have used gevent and eventlet extensively and I agree 100% that they are rock solid for async development, but nobody wants to write code that way anymore.

Yeah I know nobody does, so IMO I think they need to ask themselves why they still want collections that are transparently loaded from the database when they otherwise have decided that implicit IO is a bad idea.     Can you look around for other asyncio ORMs and maybe see what they've done here?

I see a big part of the longer term future here as continuing with the "write only" approach and adding more kinds of loaders.    this is going with the asyncio philosophy of "I should write an explicit query and execute it for all IO", we could provide for the "selectinload" loader to also be something someone could invoke explicitly and get a bunch of collections at once as a dictionary of some kind, something like that.  or the "write-only" collection can provide for an explicitly requested view (like `order.order_items.buffered_contents`, something less verbose) that gives you what's currently known to be in memory without ever trying to load.   


> 
> > you could also use asyncio with SQLAlchemy without any of these issues by using the run_sync() approach.
> 
> The solution you came up to mix sync and async code is super clever and my mind exploded when I learned it from you, but I don't think you can expect async developers will want to use `run_sync()` in their code.

yup, another cultural block created by asyncio.   fairly pointless IMO

**Reply by miguelgrinberg** [reply]
> well it would behave more consistently at least, I think the inconsistency you see is part of the problem you had.

The inconsistency I was trying to point out is related to a change in the behavior of an uninitialized relationship before and after a flush:

- If the session wasn't flushed, the relationship automatically transforms into an empty list on first use.
- After a flush, the relationship is lazy loaded on first use.
- But if I initialize the relationship to an empty list explicitly, then after a flush there is no lazy loading attempt.

More so, if I use the `write_only` loader, then no attempt to lazy load is made (obviously), but if I use `joined` or any other eager loader a lazy load is attempted anyway. I know you explained that the choice of loader has nothing to with this, but choosing `write_only` does change the lazy loading behavior, so sometimes it does. Correct me if I'm wrong, but `write_only` comes with its own customized collection class, while `joined`, `subquery`, etc. use the same container as `select`, so they all have the lazy loading risk because they share the container. This may sound like a crazy idea, but maybe in the same way we have the `WriteOnlyCollection`, we need an `EagerLoadedCollection` that has list semantics but never lazy loads. Then only `select` would use the original instrumented list.

In any case, with your explanations I now see why things happen the way they do. But looking at it from the outside it all feels somewhat arbitrary and inconsistent. And in case it isn't clear, I'm not trying to be a complainer, I really would like to figure out a way to explain this to help people avoid these obscure errors.

> Basic example, someone emits a query to load the parent object, and they explicitly do not indicate any eager loader, because they don't actually want the additional SQL. The collection will be unloaded. Later, someone edits the code that wants to access the collection, lazy="raise" and it raises. The programmer has to go to that original query and alter it.

Yes, this feels like a very good use for the `raise` loader, I have no objections. But with a brand new parent object the inconsistent behavior remains, right? Even with the `raise` loader, if I use the relationship before the session is flushed, I will get an automatic empty list and not an exception. Flushes can be issued implicitly due to an unrelated database query, so here you have a relationship that sometimes raises and sometimes doesn't when you access it for the first time. Once again, you have helped me understand how a flush changes things so I do get the difference, but given that flushes are implicit a lot of times the behavior will appear inconsistent to people who does not know the details.

> it works for new objects if you set up a default list constructor. this works if you use [declarative dataclass mappings](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping), or even just set up a custom constructor that adds a default collection. this is not a hard problem and I dont know that it's worth pages and pages of back and forth for something we are already https://github.com/sqlalchemy/sqlalchemy/discussions/9242#discussioncomment-4877302.

Yes. It does look like we don't have to discuss this anymore, since there is a solution. :)

> Literally every reply in this thread includes the workaround that was first noted in https://github.com/sqlalchemy/sqlalchemy/discussions/8913 and as you requested is [now on the live site in the docs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession), see bullet number four, it's up, with bold text....

Ah, that's great. I have missed the update to the docs. Looks good to me!

> Can you look around for other asyncio ORMs and maybe see what they've done here?

My impression is that no other ORM gets even close to the level of sophistication we get from SQLAlchemy, so this problem I'm bugging you about is unlikely to be a problem for the other ORMs. I mean, I don't think the idea of having relationships automagically work like lists is implemented anywhere else. I have been spoiled after using them for many years!

I'm going to apologize and drag this conversation to one more aspect that I think is related. Let's say we all agree that initializing the relationship to an empty list is the way to go. What happens with relationships that aren't lists? Specifically the `uselist=False` side of a 1:1 or the "one" side of a one-to-many or a many-to-one. Are those also going to have unexpected lazy loading behaviors if they are uninitialized and the session is flushed?

**Reply by zzzeek** [reply]
> > well it would behave more consistently at least, I think the inconsistency you see is part of the problem you had.
> 
> The inconsistency I was trying to point out is related to a change in the behavior of an uninitialized relationship before and after a flush:
> 
>     * If the session wasn't flushed, the relationship automatically transforms into an empty list on first use.

this is not accurate, nothing is "transformed".  the empty list is returned but it's not actually part of the object's state.   Many years ago, accessing this list **would** actually initialize the state of the object, meaning if you flushed, the behavior would be **different** if you had previously accessed the list, vs. not.   So the way it is now is actually more consistent than it used to be.

> 
>     * After a flush, the relationship is lazy loaded on first use.
> 
>     * But if I initialize the relationship to an empty list explicitly, then after a flush there is no lazy loading attempt.

IMO this is consistent.

> 
> 
> More so, if I use the `write_only` loader, then no attempt to lazy load is made (obviously), but if I use `joined` or any other eager loader a lazy load is attempted anyway. 

write-only is new and ideally this could emit a warning or error of some kind.   feel free to report a bug

> I know you explained that the choice of loader has nothing to with this, but choosing `write_only` does change the lazy loading behavior, so sometimes it does. 

"write only" is not really a lazy loader, it's an entirely different kind of collection instrumetnation, in the same way "dynamic" was


> Correct me if I'm wrong, but `write_only` comes with its own customized collection class, while `joined`, `subquery`, etc. use the same container as `select`, so they all have the lazy loading risk because they share the container. This may sound like a crazy idea, but maybe in the same way we have the `WriteOnlyCollection`, we need an `EagerLoadedCollection` that has list semantics but never lazy loads. Then only `select` would use the original instrumented list.

yes, I am proposing this in my previous comment (not that name though).


> 
> In any case, with your explanations I now see why things happen the way they do. But looking at it from the outside it all feels somewhat arbitrary and inconsistent. 

This level of complaint is way beyond what I've ever gotten for any of this behavior in 16 years so I think because you are a. using asyncio and b. trying to write a tutorial, you're in a special place with this.      please appreciate that SQLAlhcemy ORM was not written with asyncio in mind.  ORMs like SQLModel I believe were.   I'm sure they approach this problem in a better way.


> And in case it isn't clear, I'm not trying to be a complainer, I really would like to figure out a way to explain this to help people avoid these obscure errors.

for asyncio, it will be a long time before we have a bulletproof pattern that does not require special care.   My perspective is that I think it is insanely amazing SQLAlchemy can do asyncio **at all**.     

Also, I will reiterate, the popularity of asyncio in Python still seems crazy to me, that we have legions of beginner programmers using a high level scripting language, but are all jumping right in to a much more verbose, esoteric and complex execution model for interacting with network services, for mostly no reason at all because these people are not working for google.    Writing beginner tutorials for asyncio code is an uphill battle.   


> 
> Yes, this feels like a very good use for the `raise` loader, I have no objections. But with a brand new parent object the inconsistent behavior remains, right? 

I think we can stop saying this now, im not sure what level of agreement you are seeking

> My impression is that no other ORM gets even close to the level of sophistication we get from SQLAlchemy, so this problem I'm bugging you about is unlikely to be a problem for the other ORMs. I mean, I don't think the idea of having relationships automagically work like lists is implemented anywhere else. I have been spoiled after using them for many years!
> 
> I'm going to apologize and drag this conversation to one more aspect that I think is related. Let's say we all agree that initializing the relationship to an empty list is the way to go. What happens with relationships that aren't lists? Specifically the `uselist=False` side of a 1:1 or the "one" side of a one-to-many or a many-to-one. Are those also going to have unexpected lazy loading behaviors if they are uninitialized and the session is flushed?

the proposal is "default" and "default_factory" are expanded to non-dataclass use so that includes scalar attributes also

if you write your tutorial around the dataclasses feature, you might find this is all much less of a problem

**Reply by miguelgrinberg** [reply]
Okay. I think I have what I need now. I will look at the new dataclass support and see if that makes more sense to use. 

The tutorial is fairly long and is mostly about the "sync" style features of 2.0. I was hoping to show how all this stuff can be adapted to asyncio, and now I have to consider if it makes more sense to omit that and avoid all these complications and limitations.

Thank you so much for your patience!

**Reply by miguelgrinberg** [reply]
Hey, just a FYI regarding this comment you made earlier in the thread:

> SQLAlhcemy ORM was not written with asyncio in mind. ORMs like SQLModel I believe were. I'm sure they approach this problem in a better way.

I read the SQLModel docs. There is absolutely no reference to using SQLModel with asyncio, all the examples are standard functions using sync engine, sync database driver and default lazy loaded relationships, as far as I can see. It left me scratching my head in confusion. Here are the two example routes they use in the official tutorial:

```python
@app.post("/heroes/")
def create_hero(hero: Hero):
    with Session(engine) as session:
        session.add(hero)
        session.commit()
        session.refresh(hero)
        return hero

@app.get("/heroes/")
def read_heroes():
    with Session(engine) as session:
        heroes = session.exec(select(Hero)).all()
        return heroes
```

So basically, you wouldn't know this is FastAPI if you don't see how `app` is defined. FastAPI runs "sync" route handlers in a thread pool, so this is actually bypassing all the complications of async at the cost of diminished concurrency. The docs reference an "Advanced Guide" that will be written "at some point" and could include topics such as how to use the SQLAlchemy async session.

**Reply by zzzeek** [reply]
thats surprising, I thought part of the point of SQLModel was that was async.  shrugs

**Reply by CaselIT** [reply]
I think the main thing is that it uses a base that's both a declarativebase of sqlalchemy and a pydantic model.
Otherwise it's just a wrapper of sqlalchemy

**Comment by zzzeek** [comment, marked as answer]
Here is a simple mixin we can add to SQLAlchemy and advertise for use when using asyncio that will initialize all relationships

```py
from sqlalchemy import event


class InitializeRelationships:
    """initialize all relationships"""


@event.listens_for(InitializeRelationships, "init", propagate=True)
def _init_relationships(tgt, arg, kw):
    mapper = inspect(tgt.__class__)
    for arg in mapper.relationships:
        if arg.key not in kw:
            kw.setdefault(
                arg.key,
                None if not arg.uselist else arg.collection_class(),
            )

```

we can apply to your script (using all 2.0 conventions) as follows:

```py
from __future__ import annotations

import asyncio

from sqlalchemy import event
from sqlalchemy import ForeignKey
from sqlalchemy import inspect
from sqlalchemy import select
from sqlalchemy.ext.asyncio import async_sessionmaker
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship

engine = create_async_engine("sqlite+aiosqlite://", echo=True)
Session = async_sessionmaker(engine, expire_on_commit=False)


class InitializeRelationships:
    """initialize all relationships"""


@event.listens_for(InitializeRelationships, "init", propagate=True)
def _init_relationships(tgt, arg, kw):
    mapper = inspect(tgt.__class__)
    for arg in mapper.relationships:
        if arg.key not in kw:
            kw.setdefault(
                arg.key,
                None if not arg.uselist else arg.collection_class(),
            )


class Model(InitializeRelationships, DeclarativeBase):
    pass


class Order(Model):
    __tablename__ = "orders"
    id: Mapped[int] = mapped_column(primary_key=True)
    order_items: Mapped[list["OrderItem"]] = relationship(
        lazy="joined", back_populates="order"
    )


class OrderItem(Model):
    __tablename__ = "order_items"
    order_id: Mapped[int] = mapped_column(
        ForeignKey("orders.id"), primary_key=True
    )
    order: Mapped[Order] = relationship(back_populates="order_items")


async def main():
    async with engine.begin() as conn:
        await conn.run_sync(Model.metadata.create_all)
    async with Session() as session:
        async with session.begin():
            o = Order()
            session.add(o)
            await session.scalars(select(OrderItem))
            oi = OrderItem()
            session.add(oi)

            o.order_items.append(oi)


asyncio.run(main())

```

the runtime errors are gone

**Reply by miguelgrinberg** [reply]
Oh, this is very interesting. I think this is a much better option, definitely.

**Reply by miguelgrinberg** [reply]
The code as defined above breaks when you have write-only relationships in your models, as these have `collection_class = None`. The following is how I changed it so that write-only relationships are skipped.

```python
class InitializeRelationships:
    """initialize all relationships"""

@event.listens_for(InitializeRelationships, "init", propagate=True)
def _init_relationships(tgt, arg, kw):
    mapper = inspect(tgt.__class__)
    for arg in mapper.relationships:
        if arg.collection_class is None and arg.uselist:
            continue  # skip write-only and similar relationships
        if arg.key not in kw:
            kw.setdefault(
                arg.key, None if not arg.uselist else arg.collection_class())

class Model(InitializeRelationships, DeclarativeBase):
    pass
```

**Comment by CaselIT** [comment]
So I'll try to extract some actionable items here:
- add to the documentation of asyncio the snippet talked about here https://github.com/sqlalchemy/sqlalchemy/discussions/9242#discussioncomment-4886025
- support default and default_factory also in non-dataclass settings
- see if eager loaders can warn if they come across a writeonly or dynamic and arent actually going to populate anything.

anything else? Not sure @zzzeek if you want to implement the change to selectinload and write-only to have the `.buffered_contents` thing. This honestly seems like something that complicates things instead of improving them (like what happens if someone adds or removes something to it (maybe solvable by using a tuple)? Are changes to objects there persisted? etc)

maybe a better error, is it possible? I feel like we may want to get into the [`BaseException.add_note`](https://docs.python.org/3/library/exceptions.html?highlight=add_note#BaseException.add_note) thing at some point since it seems useful in situation like these

**Reply by zzzeek** [reply]
I think a mixin or mapper arg or something that does the "initialize things on construct" is where to start.

separately, see if eager loaders can warn if they come across a writeonly or dynamic and arent actually going to populate anything.  that would be a very old issue because dynamic has been around a long time.

as far as some kind of WriteOnly that also can deliver its in memory buffer, would need to think about that more.   it starts to look like we are just going to take relationship collection class and just build something new with different semantics, but it's not clear how much different it is than just plain lazy="raise".

**Reply by CaselIT** [reply]
> I think a mixin or mapper arg or something that does the "initialize things on construct" is where to start.

ah interesting, so basically use that event handler instead of modifying the init. that definitely makes sense, we just need to accept default/default_factory in non dataclass settings

I've updated the above list with the warning
